[Tf] Update TF_FOR_ALL to C++11 iterator

Cewein opened the pull request:
### Description of Change(s)

Update files under pxr/base/tf by replacing TF_FOR_ALL with C++11-style range-based for each loop.
Additionally, updated pointer usage to align with the new iterator approach, since we now handle references directly rather than raw pointers.

No modification to `iterator.h` was done nether to the part of `bits.h` and `compressedBits.h` that provide support for `TF_FOR_ALL` as it break the compilation and because the modification of the iterators regarding `bits` and `compressedBits` is out of scope of this PR.

File in sub folder have not been touch yet except `testenv/bits`.

### Ongoing Issue(s)

This PR addresses issue https://github.com/PixarAnimationStudios/OpenUSD/issues/80, though many more instances still need to be resolved. 

### Checklist

- [x] I have created this PR based on the dev branch

- [x] I have followed the [coding conventions](https://openusd.org/release/api/_page__coding__guidelines.html)

- [ ] I have added unit tests that exercise this functionality (Reference: 
[testing guidelines](https://openusd.org/release/api/_page__testing__guidelines.html))

- [x] I have verified that all unit tests pass with the proposed changes

- [x] I have submitted a signed Contributor License Agreement (Reference: 
[Contributor License Agreement instructions](https://openusd.org/release/contributing_to_usd.html#contributor-license-agreement))

Cewein committed:
update TF_FOR_ALL to C++11 iterator

As the name say.

Cewein committed:
change auto to TYPE name for clarity

When the type name was easy to write or short enough change it back to the type name rather than auto.

Long type name and namespace where excluded.

nvmkuruc commented:
Minor note-- It looks like your change description is referencing ` pxr/base/plug` instead of ` pxr/base/tf`.

Cewein commented:
> Minor note-- It looks like your change description is referencing ` pxr/base/plug` instead of ` pxr/base/tf`.

Thank you for noticing, updated the description, added also that the file under `pxr/base/tf/testenv` where almost not modified. I reserve that for another PR since it would have pilled up to too many file modified at the same time.

If you said this is alright would be happy to push the modification here.

jesschimein commented:
Filed as internal issue #USD-10653

:heavy_exclamation_mark: Please make sure that a signed CLA has been submitted!

(This is an automated message. See [here](https://openusd.org/release/contributing_to_usd.html#github-issues) for more information.)

jesschimein commented:
/AzurePipelines run

Cewein committed:
fix value acces

We already have access to the wanted value with the reference, thus no need for the * operator to access the value.

jesschimein commented:
/AzurePipelines run

meshula commented on pxr/base/tf/mallocTag.cpp:
```diff
@@ -190,8 +190,8 @@ Tf_MallocTagStringMatchTable::SetMatchList(const std::string& matchList)
     std::vector<std::string> items = TfStringTokenize(matchList, ",\t\n");
-    TF_FOR_ALL(i, items) {
-        _matchStrings.push_back(_MatchString(TfStringTrim(*i, " ")));
+    for(std::string i: items) {
```
could be a reference here, `std::string&`. could reserve `_matchStrings` to `matchList.size` even though the existing code didn't do it. `emplace_back` would allow construction in place. Another optimization that didn't originally exist, but since we're touching the loop :)

meshula commented on pxr/base/tf/errorMark.cpp:
```diff
@@ -109,11 +109,11 @@ TfReportActiveErrorMarks()
     }
 
-    TF_FOR_ALL(i, localStacks) {
+    for(const auto& i: localStacks) {
```
Throughout the diff, let's match the style of the surrounding code, with a space between for and the paren. 

In general the convention with auto is to only use it when the type is visible in the context. Like if `std::vector<float>` was declared on a previous line, `auto` would obvious from context. In the case of the _ActiveMarkStacksMap we have to go digging to discover that the iteration will be on a series of `vector<uintptr_t>` so it would help readability if the line read

`for (const vector::<uintptr_t>& i: localStacks) {`
Since the goal of this PR is to improve readability by expanding the TF_FOR_ALL macro, I think it's worthwhile to think about each replacement.

Cewein replied:
this make sense, I was not sure on how to update with `auto` or the actual type since sometime the `auto` as sometime it might translate to lengthy types.

meshula commented:
Thanks for addressing the compile issue. Since we are using this PR to figure out the best way to update the TF_FOR_ALL macros, I've left a couple of suggestions in line to help iterate ideas.

Cewein commented on pxr/base/tf/type.cpp:
```diff
@@ -1096,9 +1096,9 @@ TfType::CastFromAncestor(TfType ancestor, void* addr) const
-            if (_CastFunction *castFunc = _info->GetCastFunc(it->GetTypeid()))
+    for(const auto& it: _info->baseTypes) { 
+        if (void* tmp = it.CastFromAncestor(ancestor, addr)) {
+            if (_CastFunction *castFunc = _info->GetCastFunc(it.GetTypeid()))
```
Since recent discussion, this is a case where TF_FOR_ALL should be keep as of now in my opinion.

`it->GetTypeid()` require direct access to the iterator and will loose it when switching to range based iterator

Cewein committed:
Clean and simplify related function

I switch some function from pointer to references since the for loop allow this. Added explicite name and type for children of node.

Please do not hesite if some function should remain with TF_FOR_ALL
